std/experimental/allocator/common.d

```diff
@@ -743,3 +743,138 @@ version(unittest)
+        {
+            bool b = true;
+            static foreach (e; value)
+                b &= isAllZeroBits!(typeof(e), e);
+
+            return b;
+        }();
+    else static if (is(typeof(value) == struct) || is(typeof(value) == union))
+        enum isAllZeroBits = ()
+        {
+            bool b = true;
+            static foreach (e; value.tupleof)
+                b &= isAllZeroBits!(typeof(e), e);
+
+            return b;
+        }();
+    else
+        enum isAllZeroBits = false;
+}
+
+@nogc nothrow pure @safe unittest
+{
+    static assert(isAllZeroBits!(int, 0));
+    static assert(!isAllZeroBits!(int, 1));
+
+    import std.meta : AliasSeq;
+    foreach (Float; AliasSeq!(float, double, real))
+    {
+        assert(isAllZeroBits!(Float, 0.0));
+        assert(!isAllZeroBits!(Float, -0.0));
+        assert(!isAllZeroBits!(Float, Float.nan));
+    }
+
+    static assert(isAllZeroBits!(void*, null));
+    static assert(isAllZeroBits!(int*, null));
+    static assert(isAllZeroBits!(Object, null));
+}
+
+/+
+Is the representation of T.init known at compile time to consist of nothing but
+zero bits? Padding between a struct's fields is not considered.
++/
+package template isInitAllZeroBits(T)
+{
+    import std.traits : isStaticArray;
+    static if (isStaticArray!T)
+        static if (T.length == 0)
+            enum isInitAllZeroBits = true;
+        else
+            enum isInitAllZeroBits = isAllZeroBits!(typeof(T.init[0]), T.init[0]);
+    else
+        enum isInitAllZeroBits = isAllZeroBits!(T, T.init);
+}
+
+@nogc nothrow pure @safe unittest
+{
+    static assert(isInitAllZeroBits!(Object));
+    static assert(isInitAllZeroBits!(void*));
+    static assert(isInitAllZeroBits!uint);
+    static assert(isInitAllZeroBits!(uint[2]));
+
+    static assert(!isInitAllZeroBits!float);
+    static assert(isInitAllZeroBits!(float[0]));
+    static assert(!isInitAllZeroBits!(float[2]));
+
+    static struct S1
+    {
+        int a;
+    }
+    static assert(isInitAllZeroBits!S1);
+
+    static struct S2
+    {
+        int a = 1;
+    }
+    static assert(!isInitAllZeroBits!S2);
+
+    static struct S3
+    {
+        S1 a;
+        int b;
+    }
+    static assert(isInitAllZeroBits!S3);
+    static assert(isInitAllZeroBits!(S3[2]));
+
+    static struct S4
+    {
+        S1 a;
+        S2 b;
+    }
+    static assert(!isInitAllZeroBits!S4);
+
+    static struct S5
+    {
+        real r = 0;
+    }
+    static assert(isInitAllZeroBits!S5);
+
+    static struct S6
+    {
+
+    }
+    static assert(isInitAllZeroBits!S6);
+
+    static class C1
+    {
+        int a = 1;
+    }
+    static assert(isInitAllZeroBits!C1);
+
+    // Ensure Tuple can be read.
+    import std.typecons : Tuple;
+    static assert(isInitAllZeroBits!(Tuple!(int, int)));
+    static assert(!isInitAllZeroBits!(Tuple!(float, float)));
+
+    // Has non-zero private fields.
+    import std.random : Mt19937;
+    static assert(!isInitAllZeroBits!Mt19937);
+}
```

Comment:
It might be worthwhile to check whether this works with `const`.
[`ConstOf`](https://dlang.org/phobos/std_traits.html#ConstOf) might be helpful.